test/cmp_msg_test.c

```diff
@@ -0,0 +1,552 @@
+                                     OSSL_CMP_PKIFAILUREINFO_systemFailure, NULL);
+    fixture->err_code = -1;
+    fixture->expected = 1;      /* Expected: Message creation is successful */
+    if (!TEST_true(set1_newPkey(fixture->cmp_ctx, newkey))) {
+        tear_down(fixture);
+        fixture = NULL;
+    }
+    EXECUTE_TEST(execute_errormsg_create_test, tear_down);
+    return result;
+}
+
+
+static int test_cmp_create_pollreq(void)
+{
+    SETUP_TEST_FIXTURE(CMP_MSG_TEST_FIXTURE, set_up);
+    fixture->expected = 1;
+    EXECUTE_TEST(execute_pollreq_create_test, tear_down);
+    return result;
+}
+
+static int test_cmp_create_rr(void)
+{
+    SETUP_TEST_FIXTURE(CMP_MSG_TEST_FIXTURE, set_up);
+    fixture->expected = 1;
+    if (!TEST_true(OSSL_CMP_CTX_set1_oldCert(fixture->cmp_ctx, cert))) {
+        tear_down(fixture);
+        fixture = NULL;
+    }
+    EXECUTE_TEST(execute_rr_create_test, tear_down);
+    return result;
+}
+
+static int test_cmp_create_genm(void)
+{
+    OSSL_CMP_ITAV *itv = NULL;
+
+    SETUP_TEST_FIXTURE(CMP_MSG_TEST_FIXTURE, set_up);
+    OSSL_CMP_CTX_set_option(fixture->cmp_ctx, OSSL_CMP_OPT_UNPROTECTED_SEND, 1);
+    fixture->expected = 1;
+    itv = OSSL_CMP_ITAV_create(OBJ_nid2obj(NID_id_it_implicitConfirm), NULL);
+    if (!TEST_ptr(itv)
+            || !TEST_true(OSSL_CMP_CTX_push0_genm_ITAV(fixture->cmp_ctx, itv))) {
+        OSSL_CMP_ITAV_free(itv);
+        tear_down(fixture);
+        fixture = NULL;
+    }
+
+    EXECUTE_TEST(execute_genm_create_test, tear_down);
+    return result;
+}
+
+static int execute_certrep_create(CMP_MSG_TEST_FIXTURE *fixture) {
+    OSSL_CMP_CERTREPMESSAGE *crepmessage;
+    OSSL_CMP_CERTRESPONSE *certresp;
+    EVP_PKEY *privkey;
+    X509 *certfromresp = NULL;
+
+    crepmessage = OSSL_CMP_CERTREPMESSAGE_new();
+    certresp = OSSL_CMP_CERTRESPONSE_new();
+    ASN1_INTEGER_set(certresp->certReqId, 99);
+    certresp->certifiedKeyPair = OSSL_CMP_CERTIFIEDKEYPAIR_new();
+    certresp->certifiedKeyPair->certOrEncCert->type =
+        OSSL_CMP_CERTORENCCERT_CERTIFICATE;
+    certresp->certifiedKeyPair->certOrEncCert->value.certificate =
+        X509_dup(cert);
+    sk_OSSL_CMP_CERTRESPONSE_push(crepmessage->response, certresp);
+    if (!TEST_ptr(ossl_cmp_certrepmessage_get0_certresponse(crepmessage, 99))) {
+        return 0; /* TODO Akretisch remove { } and better avoid mem leaks */
+    };
+    if (!TEST_ptr_null(ossl_cmp_certrepmessage_get0_certresponse(
+            crepmessage, 88))) {
+        return 0; /* TODO Akretisch remove { } and better avoid mem leaks */
+    };
+    privkey = OSSL_CMP_CTX_get0_newPkey(fixture->cmp_ctx, 1);/* TODO Akretsch add check for NULL */
+    certfromresp = ossl_cmp_certresponse_get1_certificate(privkey, certresp);
+    if (certfromresp == NULL
+            || !TEST_int_eq(X509_cmp(cert, certfromresp), 0)) {
+        return 0; /* TODO Akretisch remove { } and better avoid mem leaks */
+    }
```

Comment:
Four outstanding TODOs in this function and a number of others in this file

Reply:
These TODOs should have been handled by @Akretsch before issuing this preview; I've dealt with them.